misc/logger/src/circularArray.ts

```diff
@@ -0,0 +1,65 @@
+export default class CircularArray {
+  private array: string[];
+  private head: number;
+  private currentSizeInBytes: number;
+  private capacityInBytes: number;
+
+  constructor(capacityInBytes: number) {
+    this.capacityInBytes = capacityInBytes;
+    this.currentSizeInBytes = 0;
+    this.head = 0;
+    this.array = [];
+  }
+
+  private calculateStringSize(item: string): number {
+    return new TextEncoder().encode(item).length;
+  }
+
+  public enqueue(item: string): void {
+    const itemSize = this.calculateStringSize(item);
+    if (itemSize > this.capacityInBytes) {
+      throw new Error("Item size exceeds total capacity.");
+    }
+
+    let insertIdx = this.array.length;
+    while (this.currentSizeInBytes + itemSize > this.capacityInBytes) {
+      const headItem = this.array[this.head];
+      this.currentSizeInBytes -= this.calculateStringSize(headItem);
+      insertIdx = this.head;
+      this.head = (this.head + 1) % Math.max(this.array.length, 1); // Avoid modulo 0, was causing bugs
+    }
+
+    if (insertIdx >= this.array.length) {
+      this.array.push(item);
+    } else {
+      this.array[insertIdx] = item;
+    }
```

Comment:
What's this loop doing? Looks like it's decreasing `currentSizeInBytes` until the item fits, but it's not actually removing anything from the array.

I think we should remove the items from the array until we reach our capacity threshold. And when we do this, we will need to keep track of a separate "array length" variable and use that in `get()` and `Symbol.iterator` instead of `array.length`

Reply:
When we remove decrease the size, we move `head` and then we overwrite that. You can see by the unit tests how it works. Eg, the test which goes from `a` to like `p`, the end result array only contains `m` `n` `o` `p`.  Only four entries which fit under the size max

Reply:
Can you add this unit test? It should fail as-is currently

Max capacity bytes: 4

Insert items:
- a
- b
- c
- d
- ee

Assert array contains:

- c
- d
- ee

Because when adding the `ee` it has to remove 2 entries not just 1. I see you have the `while` loop which I assume was supposed to handle removing multiple entries, but it sounds like the deletion (you call it overwrite) logic doesn't handle removing multiple entries.

Reply:
You're right, this is handled now.

Reply:
Did you add the test case?